File: 0146_LRU_Cache/main.cpp
class LRUCache {
    struct Node {
        int key;
        int value;
        Node *prev;
        Node *next;
        Node(int key, int value, Node *prev, Node *next) {
            this->key = key;
            this->value = value;
            this->prev = prev;
            this->next = next;
        }
    };
    Node *head = new Node(-1, -1, nullptr, nullptr);
    unordered_map<int, Node *> map;

    void moveTail(Node *cur) {
        if (cur == head) {
            head = head->next;
            return;
        }
        cur->prev->next = cur->next;
        cur->next->prev = cur->prev;

        cur->next = head;
        cur->prev = head->prev;
        head->prev->next = cur;
        head->prev = cur;
    }

  public:
    LRUCache(int capacity) {
        Node *cur = head;
        for (int i = 0; i < capacity - 1; i++) {
            cur->next = new Node(-1, -1, cur, nullptr);
            cur = cur->next;
        }
        cur->next = head;
        head->prev = cur;
    }

    int get(int key) {
        if (map.count(key) == 0) {
            return -1;
        }
        Node *cur = map[key];
        moveTail(cur);
        return cur->value;
    }

    void put(int key, int value) {
        if (map.count(key) != 0) {
            Node *cur = map[key];
            cur->value = value;
            moveTail(cur);
        } else {
            if (head->value != -1) {
                map.erase(head->key);
            }
            head->key = key;
            head->value = value;
            map[key] = head;
            head = head->next;
        }
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

File: 0146_LRU_Cache/golang/main.go
type node struct {
	key, value int
	prev, next *node
}

type LRUCache struct {
	head *node
	mMap map[int]*node
}

func (this *LRUCache) move2Head(n *node) {
	if n == this.head {
		this.head = this.head.prev
		return
	}
	n.next.prev = n.prev
	n.prev.next = n.next

	head := this.head
	n.next = head.next
	n.next.prev = n
	n.prev = head
	head.next = n
}

func Constructor(capacity int) LRUCache {
	lru := LRUCache{}
	lru.mMap = make(map[int]*node)
	lru.head = &node{-1, -1, nil, nil}

	p := lru.head
	for i := 1; i < capacity; i++ {
		p.next = &node{-1, -1, p, nil}
		p = p.next
	}
	p.next = lru.head
	lru.head.prev = p
	return lru
}

func (this *LRUCache) Get(key int) int {
	n, ok := this.mMap[key]
	if ok {
		this.move2Head(n)
		return n.value
	}
	return -1
}

func (this *LRUCache) Put(key int, value int) {
	n, ok := this.mMap[key]
	if ok == false {
		delete(this.mMap, this.head.key)
		this.head.key = key
		this.head.value = value
		this.mMap[key] = this.head
		this.head = this.head.prev
	} else {
		n.value = value
		this.move2Head(n)
	}
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * obj := Constructor(capacity);
 * param_1 := obj.Get(key);
 * obj.Put(key,value);
 */